lightning/src/routing/router.rs

```diff
@@ -341,21 +340,27 @@ pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, targ
 	}
 
 	for hop in last_hops.iter() {
-		if first_hops.is_none() || hop.src_node_id != *our_node_id { // first_hop overrules last_hops
-			if network.get_nodes().get(&hop.src_node_id).is_some() {
-				if first_hops.is_some() {
-					if let Some(&(ref first_hop, ref features)) = first_hop_targets.get(&hop.src_node_id) {
-						// Currently there are no channel-context features defined, so we are a
-						// bit lazy here. In the future, we should pull them out via our
-						// ChannelManager, but there's no reason to waste the space until we
-						// need them.
-						add_entry!(first_hop, *our_node_id , hop.src_node_id, dummy_directional_info, features.to_context(), 0);
-					}
-				}
-				// BOLT 11 doesn't allow inclusion of features for the last hop hints, which
-				// really sucks, cause we're gonna need that eventually.
-				add_entry!(hop.short_channel_id, hop.src_node_id, target, hop, ChannelFeatures::empty(), 0);
-			}
+		let have_hop_src_in_graph = match first_hops {
+			Some(_) => {
+				// If this hop connects to a node with which we have a direct channel, ignore the
+				// network graph and add both the hop and our direct channel to the candidate set:
+				if let Some(&(ref first_hop, ref features)) = first_hop_targets.get(&hop.src_node_id) {
+					// Currently there are no channel-context features defined, so we are a
+					// bit lazy here. In the future, we should pull them out via our
+					// ChannelManager, but there's no reason to waste the space until we
+					// need them.
+					add_entry!(first_hop, *our_node_id , hop.src_node_id, dummy_directional_info, features.to_context(), 0);
+					true
+			// In any other case, only add the hop if the source is in the regular network
+			// graph:
+				} else { network.get_nodes().get(&hop.src_node_id).is_some() }
```

Comment:
Consider using `match` here.

Reply:
Is there a particular reason to prefer match in general, or do you mean in this specific case? I find `if let` to be more readable in the two-case version because it requires less indentation in many cases.

In either case, we can simplify it further and drop the outer match, which I went ahead and did.